packages/snaps-controllers/src/interface/InterfaceController.ts

```diff
@@ -0,0 +1,219 @@
+import type { RestrictedControllerMessenger } from '@metamask/base-controller';
+import { BaseController } from '@metamask/base-controller';
+import type { Component, InterfaceState } from '@metamask/snaps-sdk';
+import { assert } from '@metamask/utils';
+import { nanoid } from 'nanoid';
+
+import { constructState } from './utils';
+
+const controllerName = 'InterfaceController';
+
+export type CreateInterface = {
+  type: `${typeof controllerName}:createInterface`;
+  handler: InterfaceController['createInterface'];
+};
+
+export type GetInterface = {
+  type: `${typeof controllerName}:getInterface`;
+  handler: InterfaceController['getInterface'];
+};
+
+export type UpdateInterface = {
+  type: `${typeof controllerName}:updateInterface`;
+  handler: InterfaceController['updateInterface'];
+};
+
+export type DeleteInterface = {
+  type: `${typeof controllerName}:deleteInterface`;
+  handler: InterfaceController['deleteInterface'];
+};
+
+export type UpdateInterfaceState = {
+  type: `${typeof controllerName}:updateInterfaceState`;
+  handler: InterfaceController['updateInterfaceState'];
+};
+
+export type InterfaceControllerActions =
+  | CreateInterface
+  | GetInterface
+  | UpdateInterface
+  | DeleteInterface
+  | UpdateInterfaceState;
+
+export type InterfaceControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  InterfaceControllerActions,
+  never,
+  InterfaceControllerActions['type'],
+  never
+>;
+
+export type StoredInterface = {
+  snapId: string;
+  content: Component;
+  state: InterfaceState;
+  usage?: string;
+};
+
+export type InterfaceControllerState = {
+  interfaces: Record<string, StoredInterface>;
+};
+
+export type InterfaceControllerArgs = {
+  messenger: InterfaceControllerMessenger;
+  state?: InterfaceControllerState;
+};
+
+/**
+ * Use this controller to manage snaps UI interfaces using RPC method hooks.
+ */
+export class InterfaceController extends BaseController<
```

Comment:
Should we call this `SnapInterfaceController`? I'm wondering if there will ever be another `InterfaceController` in a MetaMask client 😅